web/files/paginated.go

```diff
@@ -108,50 +79,73 @@ func dirData(c echo.Context, statusCode int, doc *vfs.DirDoc) error {
 		}
 	}
 	rel := jsonapi.RelationshipMap{
-		"parent":   parent,
-		"contents": jsonapi.Relationship{Data: relsData},
+		"parent": parent,
+		"contents": jsonapi.Relationship{
+			Meta: &jsonapi.RelationshipMeta{Count: count},
+			Links: &jsonapi.LinksList{
+				Self: "/files/" + doc.DocID + "/relationships/contents",
+			},
+			Data: relsData},
 	}
 
-	var links *jsonapi.LinksList
-	if hasNext && len(included) > 0 {
-		next := fmt.Sprintf("/files/%s?page[cursor]=%s&page[limit]=%d",
-			doc.DocID, included[len(included)-1].ID(), count)
-		links = &jsonapi.LinksList{Next: next}
+	var links jsonapi.LinksList
+	if cursor.HasMore() {
+		params, err := jsonapi.PaginationCursorToParams(cursor)
+		if err != nil {
+			return err
+		}
+		next := "/files/" + doc.DocID + "/relationships/contents?" + params.Encode()
+		rel["contents"].Links.Next = next
 	}
 
 	dir := &dir{
 		doc:      doc,
 		rel:      rel,
 		included: included,
 	}
-	return jsonapi.Data(c, statusCode, dir, links)
+	return jsonapi.Data(c, statusCode, dir, &links)
 }
 
 func dirDataList(c echo.Context, statusCode int, doc *vfs.DirDoc) error {
-	included := make([]jsonapi.Object, 0)
 
-	count, iterOpts, err := paginationConfig(c)
+	cursor, err := jsonapi.ExtractPaginationCursor(c, defPerPage)
+	if err != nil {
+		return err
+	}
+
+	inst := middlewares.GetInstance(c)
```

Comment:
It looks like you are just using the VFS on inst, so we can be put in a variable:

```go
fs := middlewares.GetInstance(c).VFS()
```